pep-0517.txt

```diff
@@ -503,6 +540,48 @@ package *authors*, while still allowing *end-users* to open up the
 hood and apply duct tape when necessary.
 
 
+In-Tree Build Backends
+======================
+
+In certain circumstances, projects may wish to include the source code for the
+build backend directly in the source tree, rather than referencing the backend
+via the ``requires`` key. Two specific situations where this would be expected
+are:
+
+- Backends themselves, which want to use their own features for building
+  themselves ("self-hosting backends")
+- Project-specific backends, typically consisting of a custom wrapper around a
+  standard backend, where the wrapper is too project-specific to be worth
+  distributing independently ("in-tree backends")
+
+Projects can specify that their backend code is hosted in-tree by including the
+``python-paths`` key in ``pyprject.toml``. This key contains a list of
+directories, which the frontend will make available on ``sys.path`` when
+loading the backend, and running the backend hooks. The directories in
+``python-paths`` will be added to the *start* of ``sys.path``.
+
+There are a number of restrictions on the content of the ``python-paths`` key:
+
+- Directories in ``python-paths`` are interpreted as relative to the project
+  root, and MUST resolve to the project root or a subdirectory of it.
+- The backend code MUST be loaded from one of the directories specified in
+  ``python-path`` (i.e., it is not permitted to specify ``python-path`` and
+  *not* have in-tree backend code).
+
+The first restriction is to ensure that source trees remain self-contained,
+and cannot refer to locations outside of the source tree. Frontends SHOULD
+check this condition (typically by resolving the location to an absolute path
+and checking it against the project root), and fail with an error message if
+it is violated.
+
```

Comment:
I'm personally fine with pointing out the desired usage, but not impose the check in any way. Leave it as an implementation detail.

Reply:
Practically speaking it will be more confusing if we leave it as an implementation detail, because `pip` and/or `setuptools` are going to impose this check, which means it will be the de facto standard. For major behavior like this, I think it's preferable to clearly specify it rather than make it "implementation defined behavior".

That said, it's a `SHOULD` and not a `MUST` so that does leave it implementation defined anyway.

Reply:
The point here is that projects are required to conform to the restriction. Whether the frontend checks is basically just a user friendliness question - do we give a clear error message when you violate this constraint, or do things just break randomly because you violated someone's assumption, and when you raise an issue you get told "you didn't follow the standard so tough".

I prefer it if frontends give helpful error messages early when possible, but if it's too much trouble, I'm fine with just having things break for people who don't follow the rules.

The point of the MUST restriction on projects is precisely so that it *isn't* implementation-defined, regardless of what implementations actually do ;-)